### ISSUE TITLE
"state identifier overflow" on large input list.

### ISSUE THREAD

**Issue by erikcw** [G-Research/ahocorasick_rs#62]
I'm attempting to create a large automaton (approx 2million strings) and am running into the following error:

```python

items = [...]
ac = ahocorasick_rs.AhoCorasick(items)
ValueError: state identifier overflow: failed to create state ID from 5182408960, which exceeds the max of 2147483646
```

It seems to work fine with `pyahocorasick`.

**Comment by BurntSushi** [comment]
Why not provide a reproduction?

It looks like the default implementation is a DFA, which can use a ton of space and will exhaust the state identifier (which uses `u32`) pretty quickly. You might consider trying `ContiguousNFA` or `NoncontiguousNFA` instead:

https://github.com/G-Research/ahocorasick_rs/blob/80887aae03b3863d60ecb8d09b420bd84945834c/pysrc/ahocorasick_rs/ahocorasick_rs.pyi#L4-L5

**Comment by erikcw** [comment]
Here is a reproduction (warning, each of these take quite a long time to run):

```python
import random
from ahocorasick_rs import AhoCorasick, Implementation



def make_sentence_list(sentences, sentence_length):
    """
    Produces a Sequence of "sentences" containing random words.
    """
    with open("/usr/share/dict/words") as f:
        words = f.read().splitlines()

    for n in range(sentences):
        yield " ".join(random.sample(words, k=sentence_length))

# Fails with ValueError
def test_ahocorasick_rs_large_list():
    items = list(make_sentence_list(4_000_000, 4))
    # with pytest.raises(ValueError):
    ac = AhoCorasick(items)


# Succeeds
def test_ahocorasick_rs_large_list_noncontiguous_nfa():
    items = list(make_sentence_list(4_000_000, 4))
    ac = AhoCorasick(items, implementation=Implementation.NoncontiguousNFA)

```

Does it make sense to change the type of the state identifier to a `u64` to prevent the overflow?

**Comment by erikcw** [comment]
I was just doing some additional testing with `pyahocorasick`.  The following test takes about 60 seconds to run on my machine, where `test_ahocorasick_rs_large_list_noncontiguous_nfa` took about 25 minutes.  It also consumed significantly less memory.

The README says this package should be significantly more performant on a QPS basis.  But the difference in initializing the data structures is surprising enough to lead me to believe that I'm either using it incorrectly, or there is a regression to address.

```python
import ahocorasick

def test_ahocorasick_large_list():
    ac = ahocorasick.Automaton()

    limit = 4_000_000
    for sentence in make_sentence_list(limit, 4):
        ac.add_word(sentence, sentence)
    ac.make_automaton()
    assert len(ac) == limit
    assert len(list(ac.iter(f"building a {sentence} haystack"))) == 1
```

**Comment by itamarst** [comment]
1. The default should probably be changed to the default Rust heuristic rather than hardcoding DFA, yeah (#55).
2. That example is you pasted is still DFA, rather than non-contiguous NFA, so not sure which one you're referring to re slowness.
3. I will test with that large of a number of strings.

(Various things that could cause this slowness are need to convert strings to UTF-8 for Rust, but also maybe high memory usage leading to swapping.)

Just for context, what version Python, and what OS? How much RAM does your computer have?

**Comment by itamarst** [comment]
I should probably allow passing in a iterator of a strings when constructing `AhoCorasick()`, so you can keep from having all the original Python patterns in memory at once by using a lazy iterator.

**Comment by BurntSushi** [comment]
Yeah that would be a good idea. I know my comment was big, but just to doubly clarify, towards the end I isolated the discrepancy in memory usage down to a pure Rust program without the binding involved at all, so there is definitely still a mystery here beyond "a copy of all of the patterns is being stored in memory." (The `aho-corasick` crate very specifically does not create a copy of the patterns. It never even collects them into memory. It streams the iterator provided by the user straight into trie construction.)

**Comment by BurntSushi** [comment]
OK, so I think I've mostly unraveled the mystery here. The essential difference between the non-contiguous NFA and the pyahocorasick NFA is that the former uses "vectors" to store transitions where as the latter uses a linked list. Those vectors have overhead themselves (24 bytes each, per state, on a 64-bit target) _and_ each of the vectors might actually have additional capacity inside of them in order to amortize the cost of growing those vectors as new transitions are added. This also applies to the giant `Vec<State>` as well. By the time the non-contiguous NFA is built, that `Vec<State>` might have an actual heap capacity that is ~twice as big as its actual length.

This pretty much explains the discrepancy in heap memory as recorded by `time` (peak memory) and the heap memory reported by the data structure itself. It also _likely_ explains the difference in memory usage between `pyahocorasick` and `ahocorasick_rs`.

It has definitely given me something to noodle on. Using a linked list in the non-contiguous NFA is a viable choice.

**Comment by BurntSushi** [comment]
Note that for (2), `aho-corasick 1.0.3` reduces the memory usage of the automatons, but only by making its actual heap usage at the end of construction match what `AhoCorasick::memory_usage` reports. But it won't reduce _peak_ memory usage. I have some ideas about that and I'm working on them now because it looks interesting, but I can't make any promises.

Slow DFA construction time is very much expected unfortunately. It's why it's only chosen automatically when the number of patterns is very small.

**Comment by itamarst** [comment]
I modified [the sample script above](https://github.com/G-Research/ahocorasick_rs/issues/62#issuecomment-1672214175) slightly (you don't need `list()` anymore on soon-to-be-released version since it can now accept iterables more broadly, not just lists), and ran it locally.

On my computer it now takes about 45-48 seconds both with default argument, i.e. heuristic that picks automaton, and explicit NoncontiguousNFA. Peak memory is 12GB.

From an _application_ perspective, peak memory is what matters once memory usage starts getting high enough, because that's the bottleneck that takes you from "doing fine" to "catastrophic swapping". From a library maintainer perspective where e.g. someone might create a few hundred smaller automatons at irregular intervals, ongoing non-peak memory usage of individual automatons clearly matters too. However, once we're hitting the 10s of GBs peak memory usage situation peak memory at construction matters a lot too, so thanks for looking into that @BurntSushi.

Anyway: the originally reported bug is fixed, construction time is reasonable with default parameters, so I'm going to close this and do a release.

**Comment by BurntSushi** [comment]
Just to add here, I have a [PR up on the `aho-corasick` crate](https://github.com/BurntSushi/aho-corasick/pull/121) that substantially improves things here. Peak memory usage for the non-contiguous NFA is now well below that of `pyahocorasick`. :-)

The baseline set by `pyahocorasick`:

```
(ve) [andrew@duff i62]$ python test.py write-patterns 300000 /tmp/300000.pats
(ve) [andrew@duff i62]$ python test.py write-patterns 4000000 /tmp/4000000.pats
(ve) [andrew@duff i62]$ time python test.py pyahocorasick /tmp/300000.pats

real    1.781
user    1.713
sys     0.067
maxmem  459 MB
faults  0
(ve) [andrew@duff i62]$ time python test.py pyahocorasick /tmp/4000000.pats

real    29.293
user    28.079
sys     1.196
maxmem  5699 MB
faults  0
```

Now the before with `ahocorasick_rs`:

```
(ve) [andrew@duff i62]$ time python test.py ahocorasick_rs_noncontiguous /tmp/300000.pats

real    2.477
user    2.322
sys     0.153
maxmem  1006 MB
faults  0
(ve) [andrew@duff i62]$ time python test.py ahocorasick_rs_noncontiguous /tmp/4000000.pats

real    40.730
user    38.825
sys     1.879
maxmem  12575 MB
faults  0
```

And now with my PR linked above:

```
(ve) [andrew@duff i62]$ time python test.py ahocorasick_rs_noncontiguous /tmp/300000.pats

real    1.820
user    1.752
sys     0.067
maxmem  369 MB
faults  0
(ve) [andrew@duff i62]$ time python test.py ahocorasick_rs_noncontiguous /tmp/4000000.pats

real    44.053
user    43.318
sys     0.710
maxmem  4506 MB
faults  0
```

Looks like construction time could still be improved, but it's pretty close.

**Comment by itamarst** [comment]
Thank you @BurntSushi! I will soon be releasing `ahocorasick_rs` 0.16 updated to use that.

**Comment by itamarst** [comment]
And 0.16.0 is now up on PyPI.
